src/Tasks/ManifestUtil/mansign2.cs

```diff
@@ -707,40 +767,129 @@ private static void AuthenticodeSignLicenseDom(XmlDocument licenseDom, CmiManife
+
+                        if (!Win32.CryptRetrieveTimeStamp(
+                            timeStampUrl,
+                            0,
+                            60 * 1000,  // 1 minute timeout
+                            algId,
+                            ref para,
+                            sigValueBytes,
+                            sigValueBytes.Length,
+                            ref ppTsContext,
+                            ref ppTsSigner,
+                            ref phStore))
+                        {
+                            throw new CryptographicException(Marshal.GetLastWin32Error());
+                        }
+                    }
+
+                    var timestampContext = (Win32.CRYPT_TIMESTAMP_CONTEXT)Marshal.PtrToStructure(ppTsContext, typeof(Win32.CRYPT_TIMESTAMP_CONTEXT));
+                    byte[] encodedBytes = new byte[(int)timestampContext.cbEncoded];
+                    Marshal.Copy(timestampContext.pbEncoded, encodedBytes, 0, (int)timestampContext.cbEncoded);
+                    timestamp = Convert.ToBase64String(encodedBytes);
+                }
+                finally
+                {
+                    if (ppTsContext != IntPtr.Zero)
+                        Win32.CryptMemFree(ppTsContext);
+
+                    if (ppTsSigner != IntPtr.Zero)
+                        Win32.CertFreeCertificateContext(ppTsSigner);
+
+                    if (phStore != IntPtr.Zero)
+                        Win32.CertCloseStore(phStore, 0);
+                }
+            }
+
+            return timestamp;
+        }
 
+        private static void TimestampSignedLicenseDom(XmlDocument licenseDom, string timeStampUrl, bool useSha256)
+        {
             XmlNamespaceManager nsm = new XmlNamespaceManager(licenseDom.NameTable);
             nsm.AddNamespace("r", LicenseNamespaceUri);
             nsm.AddNamespace("ds", SignedXml.XmlDsigNamespaceUrl);
             nsm.AddNamespace("as", AuthenticodeNamespaceUri);
 
-            byte[] licenseXml = Encoding.UTF8.GetBytes(licenseDom.OuterXml);
+            string timestamp = String.Empty;
 
-            unsafe
+            try
             {
-                fixed (byte* pbLicense = licenseXml)
-                {
-                    Win32.CRYPT_DATA_BLOB licenseBlob = new Win32.CRYPT_DATA_BLOB();
-                    IntPtr pvLicense = new IntPtr(pbLicense);
-                    licenseBlob.cbData = (uint)licenseXml.Length;
-                    licenseBlob.pbData = pvLicense;
+                // Try RFC3161 first
+                XmlElement signatureValueNode = licenseDom.SelectSingleNode("r:license/r:issuer/ds:Signature/ds:SignatureValue", nsm) as XmlElement;
+                string signatureValue = signatureValueNode.InnerText;
+                timestamp = ObtainRFC3161Timestamp(timeStampUrl, signatureValue, useSha256);
+            }
+            // Catch all exceptions to ensure fallback to old code (non-RFC3161)
+            catch
```

Comment:
I also don't think you should have a blanket catch.  Looks like ObtainRFC3161 will catch exceptions and only throw a CryptographicException.  

If  ObtainRFC3161 catches any other exception it would return an empty string.  You should also check timestamp for an empty string and fall back to non-RFC3161 in that case also

Reply:
This part has been keep identical to mansign2.cs from the .NET repo to keep them in sync.

ObtainRFC3161 can throw other exceptions as well besides CryptographicException. The contract is that timestamp is a valid value on function return or the function throws a exception for any erros. We should not mix up return code and exception based error handling. It is either one way or the other. In this case, if anything fails, we deal with the exception. If there is no exception, then a valid timestamp value was returned.